### PR TITLE
[mesh-forwarder] add delay-aware queue management

### PR DESCRIPTION
**Pull request by abtink**
This commit implements delay-aware queue management. When enabled the
device will monitor time-in-queue of messages in the direct tx queue
and if it is lager than specified thresholds it updates ECN flag
(if message indicates it is ECN-capable) and/or drop the message. This
mechanism is applied to IPv6 messages on the first device that sends
the message into Thread mesh and also on intermediate routers that are
forwarding the message (e.g., as a "mesh lowpan fragment" frame). On an
intermediate router when forwarding the fragments of a message, if any
fragment is dropped by the queue management policy, all subsequent
fragments will also be dropped.

In particular, this commit contains the following:

- Adds `DecompressEcn()` and `MarkCompressedEcn()` in `Lowpan` class
  to decompress or update the ECN field in a compressed IPHC header
  (unit test `test_lowpan` is also updated to test the new methods).
- Adds `UpdateEcnOrDrop()` which implements the main queue management
 logic. This method is used when preparing next direct tx message. It
 decides whether to keep the message as is, update ECN on it or drop
 it.
- Updates `EvictMessage()` to first apply the queue management rule
  to see if any message can be dropped before using the eviction
  logic based on message priority.
- Updates and reuses the `FragmentPriorityList` to track whether
  queue management dropped any of the fragments of same message so
  to also drop any subsequent ones.
- Updates `LogMessage()` to log when a message is dropped by
  queue-management or when ECN is marked on a message.

---

_Background_:
- This is related to [SPEC-1041](https://threadgroup.atlassian.net/browse/SPEC-1041).

### CODE REVIEW

**Comment by abtink** [comment]
We need to later consider how we can potentially add test-cases to cover the behavior of such features. For now I did some manual testing of the behavior:
- Added a [test branch](https://github.com/abtink/openthread/tree/testing/delay-aware-queue) where we can artificially add extra fixes delay (of 600ms) on every 5th direct tx frame
- With this change, ran some scenarios (in simulation platform) pining different size to different destination to validate the queue-managment behaviors (validating through checking the logs). I paste some of the cases below.

```
ping fe80:0:0:0:c44d:6669:43f4:fa3d 30 00:02:49.448 30 6 0.1
>>>  Sending 6 small pings with 0.1 interval in between, we see after test code 
>>> adds delay, the next message is dropped by quque-mgmy
00:02:53.909 [I] Icmp6---------: Sent echo request: (seq = 13)
00:02:53.910 [I] Mac-----------: TESING --> Adding delay before report SendDone MeshForwarder
00:02:53.910 [I] MeshForwarder-: Received IPv6 ICMP6 msg, len:78, chksum:0000, from:c64d666943f4fa3d, sec:yes, prio:normal, rss:-20.0
00:02:53.910 [I] MeshForwarder-:     src:[fe80:0:0:0:c44d:6669:43f4:fa3d]
00:02:53.910 [I] MeshForwarder-:     dst:[fe80:0:0:0:dcdf:3cb1:6525:1d90]
38 bytes from fe80:0:0:0:c44d:6669:43f4:fa3d: icmp_seq=13 hlim=64 time=1ms
00:02:54.010 [I] Icmp6---------: Sent echo request: (seq = 14)
00:02:54.114 [I] Icmp6---------: Sent echo request: (seq = 15)
00:02:54.219 [I] Icmp6---------: Sent echo request: (seq = 16)
00:02:54.323 [I] Icmp6---------: Sent echo request: (seq = 17)
00:02:54.515 [I] Mac-----------: TESING --> Reporting SendDone after delay
00:02:54.515 [I] MeshForwarder-: Sent IPv6 ICMP6 msg, len:78, chksum:0000, to:c64d666943f4fa3d, sec:yes, prio:normal
00:02:54.515 [I] MeshForwarder-:     src:[fe80:0:0:0:dcdf:3cb1:6525:1d90]
00:02:54.515 [I] MeshForwarder-:     dst:[fe80:0:0:0:c44d:6669:43f4:fa3d]
>>> Here we see the log which drops the msg.
00:02:54.515 [N] MeshForwarder-: Dropping (queue mgmt) IPv6 ICMP6 msg, len:78, chksum:0000, sec:yes, prio:normal
00:02:54.515 [N] MeshForwarder-:     src:[fe80:0:0:0:dcdf:3cb1:6525:1d90]
00:02:54.515 [N] MeshForwarder-:     dst:[fe80:0:0:0:c44d:6669:43f4:fa3d]
```

```
>>> The scenario where we mark ECN instead.
00:00:17.843 [I] Mac-----------: TESING --> Reporting SendDone after delay
00:00:17.844 [I] MeshForwarder-: Sent IPv6 ICMP6 msg, len:78, chksum:0000, to:56c2d7c07568c3ca, sec:yes, prio:normal
00:00:17.844 [I] MeshForwarder-:     src:[fe80:0:0:0:9c71:9907:e910:a75e]
00:00:17.844 [I] MeshForwarder-:     dst:[fe80:0:0:0:54c2:d7c0:7568:c3ca]
00:00:17.844 [I] MeshForwarder-: Marked ECN IPv6 ICMP6 msg, len:78, chksum:0000, sec:yes, prio:normal
00:00:17.844 [I] MeshForwarder-:     src:[fe80:0:0:0:9c71:9907:e910:a75e]
00:00:17.844 [I] MeshForwarder-:     dst:[fe80:0:0:0:54c2:d7c0:7568:c3ca]
00:00:17.848 [I] MeshForwarder-: Sent IPv6 ICMP6 msg, len:78, chksum:0000, to:56c2d7c07568c3ca, sec:yes, prio:normal
```

```
>>> Marking ECN by an intermediate router (on mesh frame).
00:00:19.349 [I] MeshForwarder-: Marked ECN mesh frame, len:47, msrc:0xa400, mdst:0x9c00, hops:2, frag:no, sec:yes
00:00:19.349 [I] MeshForwarder-:     IPv6 ICMP6 msg, chksum:0000, prio:normal
00:00:19.349 [I] MeshForwarder-:     src:[fdde:ad00:beef:0:0:ff:fe00:a400]
00:00:19.349 [I] MeshForwarder-:     dst:[fdde:ad00:beef:0:0:ff:fe00:9c00]

**Comment by EskoDijk** [comment]
I found a few instances of "enc" in the code and comments - can be changed to "ecn".

File: src/core/config/misc.h [hunk 1]
```diff
@@ -421,6 +421,68 @@
...
+ * @def OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_ENABLE
+ *
+ * Define to 1 to enable delay-aware queue management for the send queue.
+ *
+ * When enabled device will monitor time-in-queue of messages in the direct tx queue and if the wait time is lager than
+ * specified thresholds it may update ECN flag (if message indicates it is ECN-capable) or drop the message.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_ENABLE
+#define OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_ENABLE 0
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/config/misc.h]: Some of the arm-gcc CI builds on cc2538 failed with this (does not fit due to size increase).
So I changed it to enable this feature by default under Thread 1.2 or higher.

**Review by EskoDijk — COMMENTED** [review]
Just to check: there seems to be a priority-queue mechanism already implemented, with eviction based on removing a lower-priority message from the queue.

Do the Address Query and Address Notficiation messages also get an 'extra high' priority in there? (Compared to other TMF/management traffic.)   If not, should they?

**Comment by abtink** [comment]
> Do the Address Query and Address Notficiation messages also get an 'extra high' priority in there? (Compared to other TMF/management traffic.) If not, should they?

Yes  @EskoDijk . Both address query and notification (along with most of TMF and MLE msgs) use "net" priority. This level is higher than other user-exposed priorities (only for use internally by Thread for its management/control traffic).

```c++
Error AddressResolver::SendAddressQuery(const Ip6::Address &aEid)
{
    Error            error;
    Coap::Message *  message;
    Tmf::MessageInfo messageInfo(GetInstance());

    message = Get<Tmf::Agent>().NewPriorityNonConfirmablePostMessage(UriPath::kAddressQuery);
    ...
```

File: src/core/config/misc.h [hunk 2]
```diff
@@ -421,6 +421,68 @@
...
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_MARK_ECN_INTERVAL
+#define OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_MARK_ECN_INTERVAL 500
+#endif
+
+/**
+ * @OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_DROP_MSG_INTERVAL
+ *
+ * Specifies the time-in-queue threshold interval in milliseconds to drop a message if it is ECN-capable.
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/config/misc.h]: Updated the comment.

File: src/core/thread/mesh_forwarder.cpp [hunk 3]
```diff
@@ -260,6 +260,182 @@ void MeshForwarder::HandleTxDelayTimer(void)
...
+        RemoveMessageIfNoPendingTx(aMessage);
+    }
+
+    return error;
+}
+
+Error MeshForwarder::RemoveAgedMessages(void)
+{
+    // This method goes through all messages in the send queue and
+    // removes all aged messages determined based on the delay-aware
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/thread/mesh_forwarder.cpp:410]: Added this in the comment. Thanks.

**Comment by EskoDijk** [comment]
> Both address query and notification (along with most of TMF and MLE msgs) use "net" priority. 

I thought the idea was that address query/notification would get an even higher priority than other management messages, say "net++" .  That is because a pending query causes other packet(s) to be stuck in the queue so it would be most effective to prioritize the query/notification messages even above other management traffic. To avoid dropping such IP packets stuck in the queue. I'm thinking about the case a Parent resolves an address query on behalf of a Child's packet.

**Comment by abtink** [comment]
> I thought the idea was that address query/notification would get an even higher priority than other management messages, say "net++" . That is because a pending query causes other packet(s) to be stuck in the queue so it would be most effective to prioritize the query/notification messages even above other management traffic. To avoid dropping such IP packets stuck in the queue. I'm thinking about the case a Parent resolves an address query on behalf of a Child's packet.

It is possible to add another priority but I am not sure how how practical or useful it can be to have another higher priority level just for address query. In general Thread mgmnt traffic is limited so I don't expct there to be many net priority msg in queue ahead of a new address query.

**Comment by EskoDijk** [comment]
> In general Thread mgmnt traffic is limited

Not necessarily, see my remarks in the SPEC-1041 discussion (let's continue there).   So DSCP bits are available to differentiate traffic (end-to-end, locally, within the mesh).  Or if that's too hard then only the CoAP traffic for "/a/" Address Resolution API can be prioritized and not the rest.

**Review by jwhui — APPROVED** [review]

**Pull request merged**